<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Packages" data-toc-modified-id="Packages-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Packages</a></span></li><li><span><a href="#Doc" data-toc-modified-id="Doc-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Doc</a></span><ul class="toc-item"><li><span><a href="#Bayesian-model" data-toc-modified-id="Bayesian-model-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Bayesian model</a></span></li><li><span><a href="#Optimization" data-toc-modified-id="Optimization-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Optimization</a></span></li></ul></li><li><span><a href="#Functions" data-toc-modified-id="Functions-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Functions</a></span><ul class="toc-item"><li><span><a href="#Link" data-toc-modified-id="Link-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Link</a></span><ul class="toc-item"><li><span><a href="#Interface" data-toc-modified-id="Interface-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Interface</a></span></li><li><span><a href="#Logit" data-toc-modified-id="Logit-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Logit</a></span></li></ul></li><li><span><a href="#Bayesian-GLM" data-toc-modified-id="Bayesian-GLM-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Bayesian GLM</a></span><ul class="toc-item"><li><span><a href="#Interface" data-toc-modified-id="Interface-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Interface</a></span></li><li><span><a href="#BayesianLogisticRegression" data-toc-modified-id="BayesianLogisticRegression-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>BayesianLogisticRegression</a></span></li><li><span><a href="#Tests" data-toc-modified-id="Tests-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Tests</a></span></li></ul></li><li><span><a href="#Miscellenaous" data-toc-modified-id="Miscellenaous-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Miscellenaous</a></span></li></ul></li><li><span><a href="#Data" data-toc-modified-id="Data-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Data</a></span><ul class="toc-item"><li><span><a href="#Read" data-toc-modified-id="Read-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Read</a></span></li><li><span><a href="#Split-train-test" data-toc-modified-id="Split-train-test-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Split train test</a></span></li></ul></li><li><span><a href="#Studies" data-toc-modified-id="Studies-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Studies</a></span><ul class="toc-item"><li><span><a href="#Single-sigma2" data-toc-modified-id="Single-sigma2-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Single sigma2</a></span><ul class="toc-item"><li><span><a href="#Naive-model" data-toc-modified-id="Naive-model-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Naive model</a></span></li><li><span><a href="#Grid-Search" data-toc-modified-id="Grid-Search-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Grid Search</a></span><ul class="toc-item"><li><span><a href="#Optimization" data-toc-modified-id="Optimization-5.1.2.1"><span class="toc-item-num">5.1.2.1&nbsp;&nbsp;</span>Optimization</a></span></li><li><span><a href="#Test" data-toc-modified-id="Test-5.1.2.2"><span class="toc-item-num">5.1.2.2&nbsp;&nbsp;</span>Test</a></span></li></ul></li><li><span><a href="#Bayesian-optimisation" data-toc-modified-id="Bayesian-optimisation-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Bayesian optimisation</a></span><ul class="toc-item"><li><span><a href="#Optimization" data-toc-modified-id="Optimization-5.1.3.1"><span class="toc-item-num">5.1.3.1&nbsp;&nbsp;</span>Optimization</a></span></li><li><span><a href="#Retrain" data-toc-modified-id="Retrain-5.1.3.2"><span class="toc-item-num">5.1.3.2&nbsp;&nbsp;</span>Retrain</a></span></li><li><span><a href="#Test" data-toc-modified-id="Test-5.1.3.3"><span class="toc-item-num">5.1.3.3&nbsp;&nbsp;</span>Test</a></span></li></ul></li></ul></li><li><span><a href="#Multiple-sigma2" data-toc-modified-id="Multiple-sigma2-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Multiple sigma2</a></span><ul class="toc-item"><li><span><a href="#Bayesian-optimization" data-toc-modified-id="Bayesian-optimization-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Bayesian optimization</a></span><ul class="toc-item"><li><span><a href="#Optimization" data-toc-modified-id="Optimization-5.2.1.1"><span class="toc-item-num">5.2.1.1&nbsp;&nbsp;</span>Optimization</a></span></li><li><span><a href="#Retrain" data-toc-modified-id="Retrain-5.2.1.2"><span class="toc-item-num">5.2.1.2&nbsp;&nbsp;</span>Retrain</a></span></li><li><span><a href="#Test" data-toc-modified-id="Test-5.2.1.3"><span class="toc-item-num">5.2.1.3&nbsp;&nbsp;</span>Test</a></span></li></ul></li></ul></li></ul></li></ul></div>

___
# Packages

In [1]:
import os
import sys
module_dir = os.path.dirname(os.path.abspath(''))
sys.path.append(module_dir)

import numpy as np
import pandas as pd
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import  roc_auc_score, log_loss, make_scorer
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize, OptimizeResult
from scipy.special import expit
from scipy.stats import gamma

from module.bayesian_model import BayesianLogisticRegression

___
# Doc

## Bayesian model
The purpose of bayesian optimization is to model the parameters as random variables. For instance, in the use case that I am going to investigate, a binary classification, we can model the law of the dependent variable $Y_{i}$ conditioned on the parameters $\boldsymbol{\Theta}$ as a Bernoulli random variable depending on observed independent variables $\boldsymbol{X_{i}}$ and the parameters as a multivariate normal random variable with a diagonal covariance matrix (parameters of $\boldsymbol{\sigma^{2}}$, which is a vector in my notation, may differ). Vectors are written in bold format in the following equations and random variable in uppercase.
$$
\begin{align} 
Y_{i} \mid \boldsymbol{\Theta} & \sim \text{Bern}(\boldsymbol{X_{i}}^{T}.\boldsymbol{\Theta}) \\
\boldsymbol{\Theta} & \sim \mathcal{N}(0,\boldsymbol{\sigma^{2}}I)
\end{align}
$$

It is possible to get one step further by modelling the diagonal elements of the covariance matrix, $\boldsymbol{\sigma^{2}}$, as a random variable as well, $\boldsymbol{\Sigma^{2}}$, with $Y_{i}$ and $\boldsymbol{\sigma^{2}}$ conditionally independent given $\boldsymbol{\Theta}$:
$$
\begin{align} 
Y_{i} \mid \boldsymbol{\Theta} & \sim \text{Bern}(\boldsymbol{X_{i}}^{T}.\boldsymbol{\Theta}) \\
\boldsymbol{\Theta} \mid \boldsymbol{\Sigma} & \sim \mathcal{N}(0,\boldsymbol{\Sigma}) \\
\Sigma_{i}^{-1} & \sim \Gamma(\alpha_{i}, \beta_{i})
\end{align}
$$

I delibaretly changed the notation from $\boldsymbol{\sigma^{2}}$ to $\boldsymbol{\Sigma}$ to emphasize the fact that the first one is a constant vector (may be unknown) and the second one is a random variable vector.

## Optimization
I am going to focus only on the 2nd model as it is more general. The purpose is to maximize $f_{\Sigma \mid \boldsymbol{Y}}(\sigma \mid \boldsymbol{y})$, $\boldsymbol{Y}$ being the vectors of all the $Y_{i}$ random variables. Unfortunately, there is no closed form for the law of $Y_{i} \mid \boldsymbol{\Sigma}$ contrary to the normal linear case (when $Y_{i} \mid \boldsymbol{\Theta} \sim \mathcal{N}(\boldsymbol{X_{i}}^{T}.\boldsymbol{\Theta},\sigma^{2}I)$). That's being said, it is possible to approximate the density function of $\Sigma \mid \boldsymbol{Y}$ thanks to the conditional independence and Bayes theorem (1), the conditional independence between $Y$ and $\Sigma$ given $\Theta$ (3) and the conditional independence between $Y_{i}$ and $Y_{j}$ given $\Theta$ (provided that $i \ne j$):
$$
\begin{align} 
f_{\boldsymbol{\Sigma} \mid \boldsymbol{Y}}(\boldsymbol{\sigma} \mid \boldsymbol{y})
& = \frac{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} \mid \boldsymbol{\Sigma} = \sigma ) f_{\Sigma}(\sigma)
}{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} )
}
\\
& = \frac{
    f_{\boldsymbol{Y}, \boldsymbol{\Theta}, \boldsymbol{\Sigma}}(\boldsymbol{y}, \boldsymbol{\theta}, \sigma)
}{
    f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{y}, \boldsymbol{\sigma})
}
\frac{
    1
}{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} )
}
\\
& = \frac{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} \mid \boldsymbol{\Theta} = \theta )
    f_{\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{\sigma})
    f_{\boldsymbol{\Sigma}}(\boldsymbol{\sigma})
}{
    f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{y}, \boldsymbol{\sigma})
}
\frac{
    1
}{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} )
}
\\
& = \frac{
    \prod_{i=1}^{n} \mathbb{P}( Y_{i} = y_{i} \mid \boldsymbol{\Theta} = \theta )
    f_{\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{\sigma})
    f_{\boldsymbol{\Sigma}}(\boldsymbol{\sigma})
}{
    f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{y}, \boldsymbol{\sigma}) \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} )
}
\end{align}
$$

We would like to find $\boldsymbol{\sigma}$ that maximize the previous density function given the observations $\boldsymbol{y}$. The numerator can be evaluated provided that we know $\boldsymbol{\theta}$. The second element of the denominator does not depend on $\boldsymbol{\sigma}$ so we can discard it when trying to maximize our objective function. Finally, the first element of the denominator is unknown. This is why we need to approximate it by means of Laplace approximation $\tilde{f}_{\boldsymbol{\Theta} \mid Y_{i}, \boldsymbol{\sigma}}(\boldsymbol{\theta^{*}} \mid y_{i}, \boldsymbol{\sigma})$ of $f_{\boldsymbol{\Theta} \mid Y_{i}, \boldsymbol{\sigma}}(\boldsymbol{\theta^{*}} \mid y_{i}, \boldsymbol{\sigma})$:

$$
\begin{align}
L(\boldsymbol{\sigma} ; \boldsymbol{y})
&= \sum_{i=1}^{n} 
    log\left(\mathbb{P}\left( Y_{i} = y_{i} \mid \boldsymbol{\Theta} = \theta^{*} \right)\right)
    + log(f_{\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}}(\boldsymbol{\theta^{*}} \mid \boldsymbol{\sigma}))
    + log(f_{\boldsymbol{\Sigma}}(\boldsymbol{\sigma}))
    - log(\tilde{f}_{\boldsymbol{\Theta} \mid Y_{i}, \boldsymbol{\sigma}}(\boldsymbol{\theta^{*}} \mid y_{i}, \boldsymbol{\sigma}))
\end{align}
$$

## Laplace approximation
We can approximate the distibution of $\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}$ by a normal distribution with mean being the mode of the distribution of $\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}$, noted $\theta^{*}$ and its covariance matrix being $(-\nabla_{\theta}\nabla_{\theta}log(f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta^{*}} \mid \boldsymbol{y}, \boldsymbol{\sigma})))^{-1}$. Finally $\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}$ could be approximated by $\mathcal{N}(\theta^{*},(-\nabla_{\theta}\nabla_{\theta}log(f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta^{*}} \mid \boldsymbol{y}, \boldsymbol{\sigma})))^{-1})$. However, $f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta^{*}} \mid \boldsymbol{y}, \boldsymbol{\sigma})$ is not known. We can calculate it by means of the following equation:

$$
\begin{align}
f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{y}, \boldsymbol{\sigma})
&= \frac{
    f_{\boldsymbol{\Theta}, \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta}, \boldsymbol{y}, \boldsymbol{\sigma})
}{
    f_{\boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{y}, \boldsymbol{\sigma})
}
\\
& = \frac{
    \mathbb{P}( \boldsymbol{Y} = \boldsymbol{y} \mid \boldsymbol{\Theta} = \theta )
    f_{\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{\sigma})
    f_{\boldsymbol{\Sigma}}(\boldsymbol{\sigma})
}{
    f_{\boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{y}, \boldsymbol{\sigma})
}
\\
& = \frac{
    \prod_{i=1}^{n} \mathbb{P}( Y_{i} = y_{i} \mid \boldsymbol{\Theta} = \theta )
    f_{\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}}(\boldsymbol{\theta} \mid \boldsymbol{\sigma})
    f_{\boldsymbol{\Sigma}}(\boldsymbol{\sigma})
}{
    f_{\boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{y}, \boldsymbol{\sigma})
}
\end{align}
$$

Only two terms depends on $\theta$ (the product over all observed samples and the conditional density of $\boldsymbol{\Theta} \mid \boldsymbol{\Sigma}$). We know them based on our model. So we can find $\theta^{*}$ and $-\nabla_{\theta}\nabla_{\theta}log(f_{\boldsymbol{\Theta} \mid \boldsymbol{Y}, \boldsymbol{\Sigma}}(\boldsymbol{\theta^{*}} \mid \boldsymbol{y}, \boldsymbol{\sigma}))$.

## Inference
When there is a new datapoint, we would like to get the distribution:
$$
\begin{align}
\mathbb{P}( Y_{n+1} = y_{n+1} \mid \boldsymbol{Y} = \boldsymbol{y} )
&=
\int f_{ Y_{n+1}, \boldsymbol{\Sigma} \mid \boldsymbol{Y} }( y_{n+1}, \boldsymbol{\sigma} \mid \boldsymbol{y} ) \,d\boldsymbol{\sigma}
\\
&=
\int \mathbb{P}( Y_{n+1} =y_{n+1} \mid \boldsymbol{\Sigma} = \boldsymbol{\sigma}, \boldsymbol{Y} = \boldsymbol{y} ) f_{ \boldsymbol{\Sigma} \mid  \boldsymbol{Y} }( \boldsymbol{\sigma} \mid  \boldsymbol{y} ) \,d\boldsymbol{\sigma}
\\
&=
\int \int f_{ Y_{n+1}, \boldsymbol{\Theta} \mid \boldsymbol{\Sigma}, \boldsymbol{Y} }( y_{n+1}, \boldsymbol{\theta} \mid \boldsymbol{\sigma}, \boldsymbol{y} ) \, d\boldsymbol{\theta}  \, f_{ \boldsymbol{\Sigma} \mid  \boldsymbol{Y} }( \boldsymbol{\sigma} \mid  \boldsymbol{y} ) \,d\boldsymbol{\sigma}
\\
&=
\int \int \mathbb{P}( Y_{n+1} =y_{n+1} \mid \boldsymbol{\Theta} = \boldsymbol{\theta}, \boldsymbol{\theta} = \boldsymbol{\Sigma} = \boldsymbol{\sigma}, \boldsymbol{Y} = \boldsymbol{y} ) f_{ \boldsymbol{\Theta} \mid \boldsymbol{\Sigma}, \boldsymbol{Y} }( \boldsymbol{\theta} \mid \boldsymbol{\sigma}, \boldsymbol{y} ) f_{ \boldsymbol{\Sigma} \mid  \boldsymbol{Y} }( \boldsymbol{\sigma} \mid  \boldsymbol{y} ) \, d\boldsymbol{\theta} \, d\boldsymbol{\sigma}
\\
&=
\int \int \mathbb{P}( Y_{n+1} =y_{n+1} \mid \boldsymbol{\Theta} = \boldsymbol{\theta} ) f_{ \boldsymbol{\Theta} \mid \boldsymbol{\Sigma}, \boldsymbol{Y} }( \boldsymbol{\theta} \mid \boldsymbol{\sigma}, \boldsymbol{y} ) f_{ \boldsymbol{\Sigma} \mid  \boldsymbol{Y} }( \boldsymbol{\sigma} \mid  \boldsymbol{y} ) \, d\boldsymbol{\theta} \, d\boldsymbol{\sigma}
\end{align} 
$$

However, there is no closed form. So we can get an approximation with:
$$
\begin{align}
\mathbb{P}( Y_{n+1} = y_{n+1} \mid \boldsymbol{Y} = \boldsymbol{y} )
&=
\mathbb{P}( Y_{n+1} = y_{n+1} \mid \boldsymbol{\Theta} = \boldsymbol{\theta}^{*} )
\end{align} 
$$
with:
- $\boldsymbol{\theta}^{*}$ being the mode of $f_{ \boldsymbol{\Theta} \mid \boldsymbol{\Sigma}, \boldsymbol{Y} }( \boldsymbol{\theta} \mid \boldsymbol{\sigma}^{*}, \boldsymbol{y} )$ and;
- $\boldsymbol{\sigma}^{*}$ being the mode of $f_{ \boldsymbol{\Sigma} \mid  \boldsymbol{Y} }( \boldsymbol{\sigma} \mid  \boldsymbol{y} )$.

___
# Bayesian optimization

## adam

In [2]:
def adam(
    fun,
    x0,
    jac,
    args=(),
    learning_rate=0.001,
    beta1=0.9,
    beta2=0.999,
    eps=1e-8,
    startiter=0,
    maxiter=1000,
    callback=None,
    **kwargs
):
    """``scipy.optimize.minimize`` compatible implementation of ADAM -
    [http://arxiv.org/pdf/1412.6980.pdf].
    Adapted from ``autograd/misc/optimizers.py``.
    """
    x = x0
    m = np.zeros_like(x)
    v = np.zeros_like(x)

    for i in range(startiter, startiter + maxiter):
        g = jac(x, *args)

        if callback and callback(x):
            break

        m = (1 - beta1) * g + beta1 * m  # first  moment estimate.
        v = (1 - beta2) * (g**2) + beta2 * v  # second moment estimate.
        mhat = m / (1 - beta1**(i + 1))  # bias correction.
        vhat = v / (1 - beta2**(i + 1))
        x = x - learning_rate * mhat / (np.sqrt(vhat) + eps)

    i += 1
    return OptimizeResult(x=x, fun=fun(x, *args), jac=g, nit=i, nfev=i, success=True)

## loss and jac

In [3]:
def solve_single_sigma2(X, y, baysian_model, prior_dist_logpdf=None):
    def callback(intermediate_result):
        print(f"fun({intermediate_result.x})={intermediate_result.fun}")


    def loss(log_sigma2, X, y, baysian_model, prior_dist_logpdf = None):
        n_feat = X.shape[1]
        single_p = np.exp(-log_sigma2[0])
        p = np.array([single_p] * n_feat)
        m = np.array([0] * n_feat)

        res = minimize(
            baysian_model._loss,
            np.array([0] * n_feat),
            args=(X, y, m, p),
            method="BFGS",
            jac=baysian_model._jac,
        )

        theta_star = res.x

        H = baysian_model._hess(theta_star, X, y, m, p)

        out = (
            baysian_model._loss(theta_star, X, y, m, p)
            - 0.5 * n_feat * np.log(2 * np.pi)
            + 0.5 * np.linalg.slogdet(H)[1]
        )

        if prior_dist_logpdf:
            out += - prior_dist_logpdf(single_p)

        return out


    def jac(log_sigma2, X, y, baysian_model, prior_dist_logpdf = None):
        h = np.log(1 + 1e-2)
        jac_list = []
        for ii in range(len(log_sigma2)):
            xk = np.copy(log_sigma2)
            xk[ii] += h
            fk_plus_h = loss(xk, X, y, baysian_model, prior_dist_logpdf)
            xk[ii] -= 2 * h
            fk_minus_h = loss(xk, X, y, baysian_model, prior_dist_logpdf)
            jac_list.append((fk_plus_h - fk_minus_h) / 2 * h)
        return np.array(jac_list)

    return adam(
        loss,
        np.array([0.0]),
        jac,
        args=(
            X.to_numpy(),
            y.to_numpy(),
            baysian_model,
            prior_dist_logpdf
        ),
        learning_rate=0.2,
        maxiter=2000
    )

In [4]:
def get_col_to_id(cols, cats):
    out = {}
    for col in cols:
        for ii, cat in enumerate(cats):
            if isinstance(cat, str):
                if col.startswith(cat):
                    out[col] = ii
            else:
                for cc in cat:
                    if cc in col:
                        out[col] = ii
    return out

def solve_multiple_sigma2(X, y, cols, col_to_id, baysian_model, prior_dist_logpdf=None):
    def callback(intermediate_result):
        print(f"fun({intermediate_result.x})={intermediate_result.fun}")


    def loss(log_sigma2, X, y, cols, col_to_id, baysian_model, prior_dist_logpdf=None):
        n_feat = X.shape[1]
        sigma2_list = []
        for col in cols:
            sigma2_list.append(np.exp(log_sigma2[col_to_id[col]]))
        diag_sigma2 = np.array(sigma2_list)

        m = np.array([0] * n_feat)
        p = 1 / diag_sigma2

        res = minimize(
            baysian_model._loss,
            np.array([0] * n_feat),
            args=(X, y, m, p),
            method="BFGS",
            jac=baysian_model._jac,
        )

        theta_star = res.x

        H = baysian_model._hess(theta_star, X, y, m, p)

        out = (
            baysian_model._loss(theta_star, X, y, m, p)
            - 0.5 * n_feat * np.log(2 * np.pi)
            + 0.5 * np.linalg.slogdet(H)[1]
        )

        if prior_dist_logpdf:
            out += - np.sum(prior_dist_logpdf(p))

        return out


    def jac(log_sigma2, X, y, cols, col_to_id, baysian_model, prior_dist_logpdf=None):
        h = np.log(1 + 1e-2)
        jac_list = []
        for ii in range(len(log_sigma2)):
            xk = np.copy(log_sigma2)
            xk[ii] += h
            fk_plus_h = loss(xk, X, y, cols, col_to_id, baysian_model, prior_dist_logpdf)
            xk[ii] -= 2 * h
            fk_minus_h = loss(xk, X, y, cols, col_to_id, baysian_model, prior_dist_logpdf)
            jac_list.append((fk_plus_h - fk_minus_h) / 2 * h)
        return np.array(jac_list)
    
    return adam(
        loss,
        np.array([0.0]*len(set(col_to_id.values()))),
        jac,
        args=(
            X.to_numpy(),
            y.to_numpy(),
            cols,
            col_to_id,
            baysian_model,
            prior_dist_logpdf
        ),
        learning_rate=0.2,
        maxiter=2000
    )

___
# Data

## Create artificial dataset

### Classification

In [5]:
def get_classification_dataset():
    n_samples = 200
    cats = ["a", "b", "c", "d", "e", "f"]
    X = pd.DataFrame(
        data={
            "col1": np.random.choice(cats, size=n_samples),
            "col2": np.random.choice(cats, size=n_samples),
            "col3": np.random.choice(cats, size=n_samples),
        }
    )
    X_preprocessed = pd.get_dummies(X)

    theta = np.random.multivariate_normal(
        np.zeros(len(cats) * X.shape[1]),
        np.diag(np.array([1e-1] * len(cats) + [1] * len(cats) + [1e1] * len(cats))),
    )

    y = pd.Series(
        data=np.random.binomial(1, expit(np.dot(X_preprocessed.to_numpy(), theta))),
        index=X_preprocessed.index,
    )
    return X_preprocessed, y

___
# Use case: Classification case

## Seed

In [6]:
np.random.seed(1)

## Dataset

In [7]:
X, y = get_classification_dataset()
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2
)

### Naive model

In [8]:
y_pred = np.array([np.mean(y_train)] * len(y_test))

print(log_loss(y_test, y_pred))
print(roc_auc_score(y_test, y_pred))

0.5972098591974028
0.5


### Grid Search

#### Optimization

In [9]:
model = BayesianLogisticRegression()
param_grid = {"default_parameters": [{"m": 0, "p": p} for p in np.logspace(-3, 3, 25)]}
grid_search = GridSearchCV(
    model,
    param_grid,
    scoring=make_scorer(log_loss, greater_is_better=False),
    return_train_score=True,
)
grid_search.fit(X_train, y_train)
print(grid_search.best_estimator_.default_parameters["p"])

0.5623413251903491


#### Test

In [10]:
y_pred = grid_search.best_estimator_.predict(X_test)

print(log_loss(y_test, y_pred))
print(roc_auc_score(y_test, y_pred))

0.4426822775635431
0.8275862068965516


### Bayesian optimisation

In [11]:
res = solve_single_sigma2(X_train, y_train, BayesianLogisticRegression())
print(res)
print(1 / np.exp(res.x))

 success: True
     fun: 82.120627402547
       x: [ 8.796e-01]
     nit: 2000
     jac: [-5.656e-16]
    nfev: 2000
[0.41496159]


#### Retrain

In [12]:
model = BayesianLogisticRegression(default_parameters={"m": 0, "p": 1 / np.exp(res.x)})
model.fit(X_train, y_train)

BayesianLogisticRegression(default_parameters={'m': 0,
                                               'p': array([0.41496159])},
                           optimize_kwargs={}, optimize_method='L-BFGS-B',
                           prior_parameters={})

#### Test

In [13]:
y_pred = model.predict(X_test)

print(log_loss(y_test, y_pred))
print(roc_auc_score(y_test, y_pred))

0.4421421988217638
0.8307210031347962


## Multiple sigma2

In [14]:
cats = ["col1", "col2", "col3"]

In [15]:
col_to_id = get_col_to_id(X_train.columns, cats)

In [16]:
res = solve_multiple_sigma2(X_train, y_train, X_train.columns, col_to_id, BayesianLogisticRegression())
print(res)
print(1 / np.exp(res.x))

 success: True
     fun: 77.23731649541712
       x: [-9.333e+00  6.781e-01  2.017e+00]
     nit: 2000
     jac: [ 1.419e-08  8.652e-07 -1.381e-07]
    nfev: 2000
[1.13071456e+04 5.07571303e-01 1.33116278e-01]


#### Retrain

In [17]:
prior_parameters = {}

for col in X_train.columns:
    prior_parameters[col] = {"m": 0, "p": (1 / np.exp(res.x))[col_to_id[col]]}

model = BayesianLogisticRegression(prior_parameters=prior_parameters)
model.fit(X_train, y_train)

BayesianLogisticRegression(default_parameters={'m': 0, 'p': 5},
                           optimize_kwargs={}, optimize_method='L-BFGS-B',
                           prior_parameters={'col1_a': {'m': 0,
                                                        'p': np.float64(11307.145640827352)},
                                             'col1_b': {'m': 0,
                                                        'p': np.float64(11307.145640827352)},
                                             'col1_c': {'m': 0,
                                                        'p': np.float64(11307.145640827352)},
                                             'col1_d': {'m': 0,
                                                        'p': np.float64(11307.145640827352)},
                                             'col1_e': {'m': 0,
                                                        'p...
                                                        'p': np.float64(0.5075713029093174)},
                                             'col3_a': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)},
                                             'col3_b': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)},
                                             'col3_c': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)},
                                             'col3_d': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)},
                                             'col3_e': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)},
                                             'col3_f': {'m': 0,
                                                        'p': np.float64(0.13311627847392613)}})

#### Test

In [18]:
y_pred = model.predict(X_test)

print(log_loss(y_test, y_pred))
print(roc_auc_score(y_test, y_pred))

0.38824273665335945
0.8777429467084639


| Hyperoptimization method  | Log loss | AUC ROC |
| --------                  | -------  | ------- |
| naive model               | 0.60     | 0.50    |
| 1D grid search            | 0.44     | 0.83    |
| 1D bayesian optimization  | 0.44     | 0.83    |
| 3D bayesian optimizatiion | 0.39     | 0.88    |